library/train_util.py

```diff
@@ -4705,112 +4710,141 @@ def sample_images_common(
         requires_safety_checker=False,
         clip_skip=args.clip_skip,
     )
-    pipeline.to(device)
-
+    pipeline.to(distributed_state.device)
     save_dir = args.output_dir + "/sample"
     os.makedirs(save_dir, exist_ok=True)
-
+    temp_prompts = []
+    for i, prompt_dict in enumerate(prompts):
+        if isinstance(prompt_dict, str):
+            prompt_dict = line_to_prompt_dict(prompt_dict)
+        assert isinstance(prompt_dict, dict)
+        # Adds an enumerator to the dict based on prompt position. Used later to name image files. Also cleanup of extra data in original prompt dict.
+        temp_dict: dict = {}
+        temp_dict["negative_prompt"] = prompt_dict.get("negative_prompt")
+        temp_dict["sample_steps"] = prompt_dict.get("sample_steps", 30)
+        temp_dict["width"] = prompt_dict.get("width", 512)
+        temp_dict["height"] = prompt_dict.get("height", 512)
+        temp_dict["scale"] = prompt_dict.get("scale", 7.5)
+        temp_dict["seed"] = prompt_dict.get("seed")
+        temp_dict["controlnet_image"] = prompt_dict.get("controlnet_image")
+        temp_dict["prompt"]: str = prompt_dict.get("prompt", "")
+        temp_dict["sample_sampler"]: str = prompt_dict.get("sample_sampler", args.sample_sampler)
+        temp_dict["enum"] = i
+        temp_prompts.append(temp_dict)
+    prompts = temp_prompts
+    temp_prompts = []
+    num_of_processes = distributed_state.num_processes
+    # Creating list with N elements, where each element is a list of prompt_dicts, and N is the number of processess available (number of devices available)
+    # prompt_dicts are assigned to lists based on order of processes, to attempt to time the image creation time to match enum order. Probably only works when steps and sampler are identical.
+    for i in range(num_of_processes):
+        temp_prompts.append([])
+    for i in range(len(prompts)):
+        temp_prompts[i%num_of_processes].append(prompts[i])
+    prompts=temp_prompts
+    del temp_prompts
+    
     rng_state = torch.get_rng_state()
     cuda_rng_state = torch.cuda.get_rng_state() if torch.cuda.is_available() else None
 
     with torch.no_grad():
         # with accelerator.autocast():
-        for i, prompt_dict in enumerate(prompts):
-            if not accelerator.is_main_process:
-                continue
+        with distributed_state.split_between_processes(prompts) as prompt_dict_lists:
+            if accelerator.is_main_process:
+            for prompt_dict in prompt_dict_lists[0]:
```

Comment:
This is an obvious syntax error.

Instead of reindenting the entire inference into a loop, please refactor the single inference cycle from `if isinstance(prompt_dict, str):` to the trailing wandb logging into a separate function?

IOW, the result here should be something like

```python
with torch.no_grad():
     with distributed_state.split_between_processes(prompts) as prompt_dict_lists:
        for prompt_dict in prompt_dict_lists[0]:
            do_inference(prompt_dict)
```

Reply:
the line 4752 `if accelerator.is_main_process:` was a leftover line of code from when I was inserting debugging outputs while testing. I've removed it, and thanks for catching it. 

I've attempted to refactor the image inference into it's own function `sample_image_inference`. Haven't had a chance to test it out yet.